TPM simulator -add Optional macro to derive EPS seed from device hardware parameters(linux only)

mohandatla opened the pull request:
Read MAC address (6 bytes) + Disk serial number generally 10 bytes and use that as entropy when USE_DEVICE_PERSISTENT_IDENTITY is set in configure.ac
todo: By default added libudev dependency, need to modify the configure.ac to give an option to the user while building it otherwise it should just build the default DEBUG behavior.

amarochk commented:
Hi Mohan,

After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:

```C
    TPM2B_STRING(EPS_CREATION, "EPS Creation");
    DRBG_STATE           rand;
    TPM2B                deviceID;
    TPM_RC result = GetDeviceID(deviceID);
    result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
    DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
```

PRIMARY_SEED_CREATION definition should go to Global.h.

I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
![image](https://user-images.githubusercontent.com/15932886/59714999-c15b0d00-91c7-11e9-9759-4a5a17114f1a.png)

-Andrey

paulmon commented on Makefile.am:
```diff
@@ -55,7 +55,7 @@ TPMCmd_Simulator_src_tpm2_simulator_CFLAGS = $(EXTRA_CFLAGS) $(PLATFORM_INC) \
 # circular dependency beetween libplatform and libtpm
 TPMCmd_Simulator_src_tpm2_simulator_LDADD = $(libplatform) $(libtpm) \
-    $(libplatform) $(LIBCRYPTO_LIBS) $(PTHREAD_LIBS) $(LIBUDEV_LIBS)
+    $(libplatform) $(LIBCRYPTO_LIBS) $(PTHREAD_LIBS) @ADDIOTIONAL_LIBS@
```
nitpick: ADDITIONAL (spelling)

mohandatla replied:
> nitpick: ADDITIONAL (spelling)

Thanks Paul, will fix it in my next iteration.

mohandatla commented:
> deliberation

Thank you for you suggestion and offline discussion. As per your suggestion, I would make changes for EPS seed to generate using device_id.

mohandatla commented:
> Hi Mohan,
> 
> After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:
> 
> ```c
>     TPM2B_STRING(EPS_CREATION, "EPS Creation");
>     DRBG_STATE           rand;
>     TPM2B                deviceID;
>     TPM_RC result = GetDeviceID(deviceID);
>     result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
>     DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
> ```
> 
> PRIMARY_SEED_CREATION definition should go to Global.h.
> 
> I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
> 
> -Andrey

Thank you for you suggestion and offline discussion. As per your suggestion, I would make changes for EPS seed to generate using device_id. Current PR.
I would also do separate PR for PPS & SPS to be in sync to make all three seeds are in sync & persistent.  Second PR. We can discuss later about taking this up or not.

jslobodzian commented on configure.ac:
```diff
@@ -43,7 +43,19 @@ AM_INIT_AUTOMAKE([foreign subdir-objects])
+    [tpm simulator gets device persistent entropy derived from hardware. Entropy is not derived from secure source.]))
+
 PKG_CHECK_MODULES([LIBCRYPTO], [libcrypto])
+AS_IF([test "x$enable_usedevicepersistententropy" = "xyes"], [
```
>x [](start = 13, length = 1)

why x?

jslobodzian replied:
Just seems a little strange...  comparing something to "yes" which makes sense to the code reader vs.  xyes which looks like "xeyes" which is an X Windows management system utilitiy that follows your mouse cursor around.

Also...there is the bit about appending an x to the left side of $enable....  which is hard to see and looks funny

---
In reply to: [295547827](https://github.com/ms-iot/ms-tpm-20-ref/pull/1#discussion_r295547827) [](ancestors = 295547827)

mohandatla commented:
> > Hi Mohan,
> > After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:
> > ```c
> >     TPM2B_STRING(EPS_CREATION, "EPS Creation");
> >     DRBG_STATE           rand;
> >     TPM2B                deviceID;
> >     TPM_RC result = GetDeviceID(deviceID);
> >     result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
> >     DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
> > ```
> > 
> > 
> > PRIMARY_SEED_CREATION definition should go to Global.h.
> > I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
> > -Andrey
> 
> Thank you for you suggestion and offline discussion. As per your suggestion, I would make changes for EPS seed to generate using device_id. Current PR.
> I would also do separate PR for PPS & SPS to be in sync to make all three seeds are in sync & persistent. Second PR. We can discuss later about taking this up or not.



> Hi Mohan,
> 
> After some more deliberation I believe the better approach will be to deterministically seed the TPM's RNG with the DeviceID _only_ while generating the Endorsement Primary Seeds (EPS) in [HierarchyPreInstall_Init()](https://github.com/microsoft/ms-tpm-20-ref/blob/9b158f43484b111bf3f6c708b436102c3e71dada/TPMCmd/tpm/src/subsystem/Hierarchy.c#L62). To do this you can use the following scheme:
> 
> ```c
>     TPM2B_STRING(EPS_CREATION, "EPS Creation");
>     DRBG_STATE           rand;
>     TPM2B                deviceID;
>     TPM_RC result = GetDeviceID(deviceID);
>     result = DRBG_InstantiateSeeded(&rand, &deviceID, EPS_CREATION, NULL, NULL);
>     DRBG_Generate(&rand, (BYTE*)gp.EPSeed.t.buffer, gp.EPSeed.t.size);
> ```
> 
> PRIMARY_SEED_CREATION definition should go to Global.h.
> 
> I'd also suggest to rename GetDeviceEntropy() to GetDeviceID(), DeviceEntropy.c o DeviceID.c, and USE_DEVICE_PERSISTENT_IDENTITY to USE_DEVICE_ID.
> 
> -Andrey

Andrey, updated PR to derive the EPS seed from device id. Added code to derive the PPS & SPS from device id as well. I can make the PPS & SPS seed derivation as a separate optional feature in case if you think is totally against the TCG standards or other reason.

mohandatla commented on TPMCmd/tpm/src/subsystem/Hierarchy.c:
```diff
@@ -64,6 +64,7 @@ HierarchyPreInstall_Init(
 #else
     CryptRandomGenerate(gp.EPSeed.t.size, gp.EPSeed.t.buffer);
 #endif
+
```
somehow missed, will do in my final interation.
